File: 2022/07/06/Q155_Min_Stack.py
class MinStack:

    def __init__(self):
        # Construct two stack;
        self.stack = []
        # Everytime the main stack gets pushed or poped, min stack do the same thing.
        self.minStack = []
        

    def push(self, val: int) -> None:
        self.stack.append(val)
        if(len(self.minStack)!=0):
            #Append into the stack the minimum of either the stack or the val.
            self.minStack.append(min(val,self.minStack[-1]))
        else:
            #Since the stack is empty.
            self.minStack.append(val)

    def pop(self) -> None:
        #Pop them both at the same time.
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        #return the very last value on the stack.
        return self.stack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()

File: 2022/07/07/Stella/Q155_Min_Stack.cpp
// Design a stack that supports push, pop, top, and retrieving the minimum element in constant time.

// Implement the MinStack class:

// MinStack() initializes the stack object.
// void push(int val) pushes the element val onto the stack.
// void pop() removes the element on the top of the stack.
// int top() gets the top element of the stack.
// int getMin() retrieves the minimum element in the stack.
// You must implement a solution with O(1) time complexity for each function.

#include <stack>

class MinStack {
public:
    
    stack<int> num;
    stack<int> min;
    
    MinStack() {
        
    }
    
    void push(int val) {
        num.push(val);
        if(min.size()==0){
            min.push(val);
        }
        else if(val<=min.top()){
            min.push(val);
        }
    }
    
    void pop() {
        if(num.top()==min.top()){
            min.pop();
        }
        num.pop();
    }
    
    int top() {
        return num.top();
    }
    
    int getMin() {
        return min.top();
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */